charts/secretsync/templates/_helpers.tpl

```diff
@@ -0,0 +1,95 @@
+{{/*
+Generate subjects for role binding and cluster role binding.
+*/}}
+{{- define "secret-sync-controller.subjects" -}}
+- kind: ServiceAccount
+  name: "secret-sync-controller-manager"
+  namespace: {{ .Values.namespace }}
+{{- end }}
+
+{{/* Generate match condition expression */}}
+{{- define "chartname.matchConditionExpression" -}}
+{{- printf "request.userInfo.username == 'system:serviceaccount:%s:%s'" .Values.namespace .Values.controllerName -}}
+{{- end -}}
+
+{{/*
+Generate allowed secret types list as a complete expression.
+*/}}
+{{- define "chartname.secretTypesList" -}}
+{{- $secretTypes := . -}}
+{{- if not $secretTypes -}}
+false
+{{- else -}}
+(object.type in [{{ range $index, $type := $secretTypes }}{{ if $index }}, {{ end }}"{{ $type }}"{{ end }}])
+{{- end -}}
+{{- end -}}
+
+{{/* Define a constant value for labelKey */}}
+{{- define "secret-sync-controller.labelKey" -}}
+secrets-store.sync.x-k8s.io
+{{- end -}}
+
+{{/* Define a constant value for labelValue */}}
+{{- define "secret-sync-controller.labelValue" -}}
+''
+{{- end -}}
+
+{{/*
+Check if the old secret has the expected label key.
+*/}}
+{{- define "secret-sync-controller.oldSecretHasExpectedLabelKey" -}}
+variables.oldSecretHasLabels && ('{{ include "secret-sync-controller.labelKey" . }}' in oldObject.metadata.labels) ? true : false
+{{- end -}}
+
+{{/*
+Check if the old secret has the expected label value.
+*/}}
+{{- define "secret-sync-controller.oldSecretHasExpectedLabelValue" -}}
+{{ include "secret-sync-controller.labelValue" . }} == oldObject.metadata.labels['{{ include "secret-sync-controller.labelKey" . }}'] ? true : false
+{{- end -}}
+
+
+{{/*
+Generate token audience comparison expression.
+Returns 'false' if tokenRequestAudience list is empty.
+*/}}
+{{- define "secret-sync-controller.tokenAudienceComparison" -}}
+{{- $tokenAudiences := .Values.tokenRequestAudience -}}
+{{- if not $tokenAudiences -}}
+false
+{{- else -}}
+{{- $audienceExpressions := list -}}
+{{- range $index, $audience := $tokenAudiences }}
+  {{- $expressionPart := printf "object.spec.audiences.exists(w, w == '%s')" $audience.audience -}}
+  {{- $audienceExpressions = append $audienceExpressions $expressionPart -}}
+{{- end -}}
+{{- join " || " $audienceExpressions -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Generate a comma-separated string from a list.
+*/}}
+{{- define "secret-sync-controller.listToString" -}}
+{{- $tokenRequests := .Values.tokenRequestAudience -}}
+{{- $audiences := list -}}
+{{- range $index, $request := $tokenRequests }}
+  {{- $audiences = append $audiences $request.audience -}}
+{{- end -}}
+{{- join ", " $audiences -}}
+{{- end -}}
+
+{{/*
+Determine the api version for the validating admission policies.
+*/}}
+{{- define "secret-sync-controller.admissionApiVersion" -}}
+{{- if semverCompare "~1.27.0" .Values.validatingAdmissionPolicies.kubernetesReleaseVersion -}}
```

Comment:
~1.x.y will only match versions where the patch version increases. We should change this block to use >= and flip the cases so >= 1.29.0-0 comes first and then >=1.28.0-0 and so on. Then we can fall through the cases we want to catch and anything greater than 1.29 will not be classed as `unsupported-validating-admission-api-version`.

Reply:
We are using tilde (~) for version compare. It means any version that is compatible with for eg. 1.29.0. Hence in semverCompare any version that is greater than or equal to 1.29.0 and less than 2.0.0 is considered compatible.